### PR TITLE
[JTC] Reject trajectories with nonzero terminal velocity

### PR DESCRIPTION
**Pull request by christophfroehlich**
As suggested by @bmagyar, the new trajectory should be even ignored if the last velocity point is nonzero. 
- This is valid for both interfaces (topic + actions).
- I think there might be use-cases, where this is intentionally (maybe receiving topics from a MPC with moving horizon, where the end of the trajectory is not a steady state). Therefore, I added a parameter to be able to deactivate this behavior. To avoid breaking changes, I deactivated this by default now but give a deprecation warning if `allow_nonzero_velocity_stop==true`.

I also added a test for that, please review #603 first.

### CODE REVIEW

**Comment by destogl** [comment]
> As suggested by @bmagyar, the new trajectory should be even ignored if the last velocity point is nonzero.

I don't agree with this at all. If we have velocity only input, we want to continue moving between trajectories. This is not so unusual scenario, e.g., when we have joystick input. I would rather like to see a timeout if we have velocity-only input to stop the input.

**Comment by christophfroehlich** [comment]
Ok, I did not think of using JTC with Joystick, but rejecting it always did not seem to make sense for me too.
So you mean we should even not hold the position automatically but only after some timeout reached?

**Comment by destogl** [comment]
> So you mean we should even not hold the position automatically but only after some timeout reached?

Exactly, I think we want to avoid run away if only velocity is used.

**Comment by christophfroehlich** [comment]
now I'm confused :D maybe we should discuss this in the next WG meeting
I think there are different use-cases, and you have others in mind than I do..

**Comment by BarisYazici** [comment]
any update merging this PR?

**Comment by christophfroehlich** [comment]
We recently decided to merge several smaller changes (#567 #558 and some optional time-out for topic interface) for JTC into a feature branch, test them thoroughly together, and merge them then together into master. 
I fear that this will take some more weeks to happen.

**Comment by christophfroehlich** [comment]
@BarisYazici you could help out by reviewing (and hopefully approving) this PR. I think everyone can do this without special permissions (goto Files changed -> Review Changes on top right corner)

**Comment by christophfroehlich** [comment]
I've split this PR into two, the more important one is #608

**Comment by christophfroehlich** [comment]
the order listed in #607 would be fine

File: joint_trajectory_controller/doc/userdoc.rst [hunk 1]
```diff
@@ -175,6 +175,9 @@ open_loop_control (boolean)
 
   Default: false
 
+allow_nonzero_velocity_stop (boolean)
```
Comment (destogl) [inline joint_trajectory_controller/doc/userdoc.rst:178]: Proposal:
```suggestion
allow_nonzero_velocity_at_trajectory_end (boolean)
```

Current name sounds to me to have inverse logic, then the description says.

Comment (christophfroehlich) [reply joint_trajectory_controller/doc/userdoc.rst]: You are right -> I pushed a commit renaming the parameter.

**Review by saikishor — APPROVED** [review]

**Comment by christophfroehlich** [comment]
The failing test is not related to this PR, but shows the flakiness of the JTC tests 😕

File: joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml [hunk 2]
```diff
@@ -66,6 +66,11 @@ joint_trajectory_controller:
       one_of<>: [["splines", "none"]],
     }
   }
+  allow_nonzero_velocity_at_trajectory_end: {
+    type: bool,
+    default_value: true,
+    description: "If not set true, the last velocity point has to be zero or the goal will be rejected",
```
Comment (bmagyar) [inline joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml:72]:
```suggestion
    description: "If false, the last velocity point has to be zero or the goal will be rejected",
```

File: joint_trajectory_controller/doc/userdoc.rst [hunk 3]
```diff
@@ -175,6 +175,11 @@ open_loop_control (boolean)
 
   Default: false
 
+allow_nonzero_velocity_at_trajectory_end (boolean)
+  If not set true, the last velocity point has to be zero or the goal will be rejected.
```
Comment (bmagyar) [inline joint_trajectory_controller/doc/userdoc.rst:179]:
```suggestion
  If false, the last velocity point has to be zero or the goal will be rejected.
```

**Pull request merged**